### PR TITLE
Implement dependency injection for Actors

### PR DESCRIPTION
**Pull request by rynowak**
# Description

This change allows Actors to accept dependency-injected services through
constructor parameters.

There are numerous breaking changes in this PR that are visible to user
code. In most cases these are simplifications.

For example the Actor base class accepted 3 constructor parameters,
which meant that every subclass (every actor type) also had 3 constructor
 parameters. This isn't fun and becomes way less fun now that users can
add their own constructor parameters in a meaningful way.

As another example, we had multiple places where "runtime options" lived
that had different interaction patterns. Now there's a single options
type and you interactive with it through `UseActors(...)`.

## Issue reference

Fixes: #171

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Extended the documentation

### CODE REVIEW

**Comment by rynowak** [comment]
This is not ready for merge yet because I haven't added tests. I wanted to get this out before it was totally ready for early feedback and so that I can see a code coverage diff - I think that very little of the functionality I'm changing has tests.

**Comment by rynowak** [comment]
@vinayada1 - ready for review

Integration tests! Wow 🎆

File: src/Dapr.Actors/Runtime/ActorManager.cs [hunk 1]
```diff
@@ -199,25 +202,86 @@ async Task<byte[]> RequestFunc(Actor actor, CancellationToken ct)
...
-            // Always add the new instance.
-            this.activeActors.AddOrUpdate(actorId, actor, (key, oldValue) => actor);
+            //
+            // In theory the Dapr runtime protects us from double-activation - there's no case
+            // where we *expect* to see the *update* code path taken. However it's a possiblity and
+            // we should handle it.
+            //
+            // The policy we have chosen is to always keep the registered instance if we hit a double-activation
+            // so that means we have to destroy the 'new' instance.
+            var current = this.activeActors.AddOrUpdate(actorId, state, (key, oldValue) => oldValue);
```
Comment (vinayada1) [inline src/Dapr.Actors/Runtime/ActorManager.cs:229]: Why don't we check if the actor is active and activate it/add it only if it does not exist?

File: src/Dapr.Actors/Runtime/ActorRuntimeOptions.cs [hunk 2]
```diff
@@ -3,29 +3,117 @@
 // Licensed under the MIT License.
 // ------------------------------------------------------------
 
+using System;
```
Comment (vinayada1) [inline src/Dapr.Actors/Runtime/ActorRuntimeOptions.cs:6]: nit: inside namespace

File: samples/Actor/DemoActor/DemoActor.cs [hunk 3]
```diff
@@ -17,18 +16,22 @@ namespace DaprDemoActor
...
+        private readonly BankService bank;
+
         /// <summary>
         /// Initializes a new instance of the <see cref="DemoActor"/> class.
         /// </summary>
         /// <param name="service">Actor Service hosting the actor.</param>
-        /// <param name="actorId">Actor Id.</param>
-        public DemoActor(ActorService service, ActorId actorId)
-            : base(service, actorId)
+        public DemoActor(ActorHost service, BankService bank)
```
Comment (vinayada1) [inline samples/Actor/DemoActor/DemoActor.cs:29]: add param name comment for bank

Comment (vinayada1) [reply samples/Actor/DemoActor/DemoActor.cs:29]: rename service to actorHost

File: src/Dapr.Actors/Runtime/ActorManager.cs [hunk 4]
```diff
@@ -199,25 +202,86 @@ async Task<byte[]> RequestFunc(Actor actor, CancellationToken ct)
...
+
+            // On this code path it was an *Update*. We need to destroy the new instance and clean up.
+            await DeactivateActorCore(state);
         }
 
-        internal async Task DeactivateActor(ActorId actorId)
+        private async Task<ActorActivatorState> CreateActorAsync(ActorId actorId)
+        {
+            this.logger.LogDebug("Creating Actor of type {ActorType} with ActorId {ActorId}", this.ActorTypeInfo.ImplementationType, actorId);
+            var actorService = new ActorHost(this.ActorTypeInfo, actorId, this.loggerFactory);
```
Comment (vinayada1) [inline src/Dapr.Actors/Runtime/ActorManager.cs:243]: rename actorService?

File: src/Dapr.Actors/Runtime/ActorManager.cs [hunk 5]
```diff
@@ -199,25 +202,86 @@ async Task<byte[]> RequestFunc(Actor actor, CancellationToken ct)
...
+                // Ensure we don't leak resources if user-code throws during deactivation.
+                await DeleteActorAsync(state);
+            }
+        }
+
+        private async ValueTask DeleteActorAsync(ActorActivatorState state)
+        {
+            this.logger.LogDebug("Deleting Actor of type {ActorType} with ActorId {ActorId}", this.ActorTypeInfo.ImplementationType, state.Actor.Id);
+            await this.activator.DeleteAsync(state);
+            this.logger.LogDebug("Finished creating Actor of type {ActorType} with ActorId {ActorId}", this.ActorTypeInfo.ImplementationType, state.Actor.Id);
```
[the remark this answers is not in the document]
Comment (rynowak) [reply src/Dapr.Actors/Runtime/ActorManager.cs:274]: What needs to be fixed?

File: src/Dapr.Actors/Runtime/ActorManager.cs (hunk 1 above)
Comment (rynowak) [reply src/Dapr.Actors/Runtime/ActorManager.cs:229]: That approach fundamentally has a race condition if the existing actor is being deactivated at the same time. Concurrent dictionary doesn't provide the guarantee that you can do that atomically.

File: samples/Actor/DemoActor/DemoActor.cs (hunk 3 above)
Comment (rynowak) [reply samples/Actor/DemoActor/DemoActor.cs:29]: We should stop adding doc-comments on samples.

They are useful when someone is going to inspect the code in the IDE (libraries) or when the generated documentation is going to be uploaded somewhere. For sample code it's just busywork. Most doc-comments are boilerplate and say nothing of substance. If we have something to say about a sample we should do with a normal comment where it's more readable.

Will rename the parameter.

File: src/Dapr.Actors/Runtime/ActorManager.cs (hunk 4 above)
Comment (rynowak) [reply src/Dapr.Actors/Runtime/ActorManager.cs:243]: will do a global search :)

File: test/Dapr.Actors.Test/Runtime/ActorTests.cs [hunk 6]
```diff
@@ -53,12 +52,10 @@ public void TestResetStateAsync()
         private TestActor CreateTestDemoActor(IActorStateManager actorStateManager)
         {
             var actorTypeInformation = ActorTypeInformation.Get(typeof(TestActor));
-            Func<ActorService, ActorId, TestActor> actorFactory = (service, id) =>
-                new TestActor(service, id, actorStateManager);
             var loggerFactory = new LoggerFactory();
-            var actorService = new ActorService(actorTypeInformation, loggerFactory, actorFactory);
-            var testActor = actorFactory.Invoke(actorService, ActorId.CreateRandom());
+            var actorService = new ActorHost(actorTypeInformation, ActorId.CreateRandom(), loggerFactory);
```
Comment (vinayada1) [inline test/Dapr.Actors.Test/Runtime/ActorTests.cs:56]: rename actorService

File: test/Dapr.Actors.Test/ITestActor.cs [hunk 7]
```diff
@@ -35,9 +35,13 @@ public interface ITestActor : IActor
     /// </summary>
     public class TestActor : Actor,  ITestActor
     {
-        public TestActor(ActorService actorService, ActorId actorId, IActorStateManager actorStateManager)
-            : base(actorService, actorId, actorStateManager)
+        public TestActor(ActorHost actorService, IActorStateManager stateManager = null)
```
Comment (vinayada1) [inline test/Dapr.Actors.Test/ITestActor.cs:38]: rename actorService

File: samples/Actor/DemoActor/DemoActor.cs (hunk 3 above)
Comment (vinayada1) [reply samples/Actor/DemoActor/DemoActor.cs:29]: where is this constructor used?

File: test/Dapr.Actors.Test/Runtime/ActorRuntimeOptionsTests.cs [hunk 8]
```diff
@@ -5,38 +5,36 @@
...
     {
         [Fact]
-        public void TestRegisterActor_SavesActorServiceFactory()
+        public void TestRegisterActor_SavesActivator()
         {
             var actorType = typeof(TestActor);
             var actorTypeInformation = ActorTypeInformation.Get(actorType);
-            var actorService = new ActorService(actorTypeInformation, new LoggerFactory());
-            Func<ActorTypeInformation, ActorService> actorServiceFactory = (actorTypeInfo) => NewActorServiceFactory(actorTypeInfo);
+            var actorService = new ActorHost(actorTypeInformation, ActorId.CreateRandom(), new LoggerFactory());
```
Comment (vinayada1) [inline test/Dapr.Actors.Test/Runtime/ActorRuntimeOptionsTests.cs:20]: rename actorService

File: test/Dapr.Actors.Test/Runtime/ActorManagerTests.cs [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+            Assert.True(Assert.IsType<TestActor>(actor).IsDeactivated);
+            Assert.True(Assert.IsType<TestActor>(actor).IsDisposed);
+        }
+
+        [Fact]
+        public async Task DeactivateActorAsync_ItsOkToDeactivateNonExistentActor()
+        {
+            var manager = CreateActorManager(typeof(TestActor));
+
+            var id = ActorId.CreateRandom();
```
Comment (vinayada1) [inline test/Dapr.Actors.Test/Runtime/ActorManagerTests.cs:129]: Assert.False(manager.TryGetActorAsync  before deactivate

File: src/Dapr.Actors/Runtime/ActorRuntimeOptions.cs (hunk 2 above)
Comment (rynowak) [reply src/Dapr.Actors/Runtime/ActorRuntimeOptions.cs:6]: I'm going to log an issue to change this convention - it's not the default that any of the tools use.

Comment (rynowak) [reply src/Dapr.Actors/Runtime/ActorRuntimeOptions.cs:6]: https://github.com/dapr/dotnet-sdk/issues/489

**Review by vinayada1 — APPROVED** [review]

**Review by vinayada1 — APPROVED** [review]

**Pull request merged**